Immutable local/remote SocketAddress within a ConnectionMetaData

gregw opened the pull request:
The local/remote SocketAddress is cached within the ConnectionMetaData or Connection instance, so that any changes are not visible during the request lifetime. 
Ensure that all server Connection types respect HttpConfiguration#getLocalAddress

This is an alternate to both #10861 and #10815 to fix #10806

Co-authored-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

gregw committed:
Immutable local/remote SocketAddress within a ConnectionMetaData

The local/remote SocketAddress is cached within the ConnectionMetaData or Connection instance, so that any changes are not visible during the request lifetime.
Ensure that all server Connection types respect HttpConfiguration#getLocalAddress

gregw commented:
@joakime Can you remind me why `HttpConfiguration#getLocalAddress` was not implemented as a request Customizer?

Could we change the implementation so that if a non null address is set on the HttpConfiguration, then that is equivalent to adding a customizer that will force the address.
We can then remove all the handling in the connections for this.

gregw committed:
Immutable local/remote SocketAddress within a ConnectionMetaData

The local/remote SocketAddress is cached within the ConnectionMetaData or Connection instance, so that any changes are not visible during the request lifetime.
Ensure that all server Connection types respect HttpConfiguration#getLocalAddress

joakime committed:
Merge remote-tracking branch 'origin/jetty-12.0.x' into fix/jetty-12.0.x/immutableAddressesInConnectionMetaData

joakime commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/SelectableChannelEndPoint.java:
```diff
@@ -133,7 +133,7 @@ public SocketAddress getLocalSocketAddress()
             if (channel instanceof NetworkChannel)
                 return ((NetworkChannel)channel).getLocalAddress();
-            return super.getLocalSocketAddress();
+            return null;
```
is returning `null` a wise thing here?
Wouldn't the HttpConnection.localAddress be appropriate here?

gregw replied:
The super implementation was just to return null.  I removed the default implementation to ensure it was clear that implementations needed to provide this.
So returning null is not new.

joakime commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/DatagramChannelEndPoint.java:
```diff
@@ -46,6 +46,21 @@ public DatagramChannel getChannel()
+            if (LOG.isTraceEnabled())
+                LOG.trace("ignored", e);
+        }
+        return null;
```
Can we return HttpConfiguration.remoteAddress here? (not sure, as this is deep in jetty-io)

gregw replied:
No need to.  We are applying HttpConfiguration.localAddress at the level of the Connection, not the endpoint.   There is not HttpConfiguration.remoteAddress

joakime commented on jetty-core/jetty-http2/jetty-http2-server/src/main/java/org/eclipse/jetty/http2/server/internal/HTTP2ServerConnection.java:
```diff
@@ -74,6 +76,8 @@ public HTTP2ServerConnection(Connector connector, EndPoint endPoint, HttpConfigu
         this.listener = listener;
         this.httpConfig = httpConfig;
         this.id = StringUtil.randomAlphaNumeric(16);
+        localSocketAddress = httpConfig.getLocalAddress() != null ? httpConfig.getLocalAddress() : endPoint.getLocalSocketAddress();
+        remoteSocketAddress = endPoint.getRemoteSocketAddress();
```
Seems like we can have a null localSocketAddress and remoteSocketAddress this way, is that wise?

gregw replied:
Always could.   If there is no address, there is no address.

joakime reviewed (commented):
Just questions about the returning of null and storing of null in final fields.
And seemingly missing handling of HttpConfiguration.(local)Address.

gregw committed:
Immutable local/remote SocketAddress within a ConnectionMetaData

The local/remote SocketAddress is cached within the ConnectionMetaData or Connection instance, so that any changes are not visible during the request lifetime.
Ensure that all server Connection types respect HttpConfiguration#getLocalAddress and that it is not implemented only in servlet layer
Avoid DNS resolution.

gregw commented:
@joakime I've pushed a few more commits dealing with HttpConfiguration.getLocalAddress().   We were dealing with that in servlet land... which was fine when we didn't have a core request.  Now that it is dealt with at the connection level, we can remove all that handling from the servlet implementation(s).

sbordet commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/AbstractMetaDataConnection.java:
```diff
@@ -0,0 +1,72 @@
+    private final SocketAddress _localSocketAddress;
+    private final SocketAddress _remoteSocketAddress;
+
+    public AbstractMetaDataConnection(Connector connector, EndPoint endPoint, Executor executor, HttpConfiguration httpConfiguration)
```
No need to pass the `Executor` as it is always taken from the `Connector`.
Also, I'd like to have longer lifecycle components first, so `(Connector, HttpConfiguration, EndPoint)` would be best.

sbordet commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/DatagramChannelEndPoint.java:
```diff
@@ -46,6 +46,21 @@ public DatagramChannel getChannel()
     }
 
+    @Override
+    public SocketAddress getRemoteSocketAddress()
```
This should be moved to `SelectableChannelEndPoint` where `getLocalSocketAddress()` is.
And be removed also from `SocketChannelEndPoint`.

gregw replied:
You can't as the `getRemoteAddress` on a `DatagramChannel` is a different method than the one on `SocketChannel` and does not exist in `NetworkChannel`.

There is a reason it is as it is.

gregw committed:
Immutable local/remote SocketAddress within a ConnectionMetaData

The local/remote SocketAddress is cached within the ConnectionMetaData or Connection instance, so that any changes are not visible during the request lifetime.
Ensure that all server Connection types respect HttpConfiguration#getLocalAddress and that it is not implemented only in servlet layer
Avoid DNS resolution.
